OpenRA.Game/Traits/Player/Shroud.cs

```diff
@@ -156,11 +159,19 @@ void INotifyCreated.Created(Actor self)
 				self.World.AddFrameEndTask(w => ExploreAll());
 		}
 
-		void Invalidate(IEnumerable<PPos> changed)
+		void ITick.Tick(Actor self)
 		{
-			foreach (var puv in changed)
+			if (OnShroudChanged == null)
+				return;
+
+			foreach (var puv in map.ProjectedCellBounds)
 			{
 				var uv = (MPos)puv;
+				if (!touched[uv])
+					continue;
```

Comment:
Looping over the entire map grid, even when no shroud needs updating, makes this the most expensive trait to tick at the moment!

Since in the usual case we don't expect many cells to change per tick (as a fraction of the total number of cells on the map) we should probably optimize with that in mind.

Consider:
- A HashSet. You only have to enumerate the changed cells, rather than all cells.
- Using a coarse array of touched regions (i.e. the dirty bins part that's now gone from FrozenActorLayer`). You track regions of say 10x10 cells in a single array entry, and when you touch any cells, you also touch the region it belongs too. In Tick, you enumerate the coarse bins instead. Any untouched bins can be skipped entirely, saving you enumerating a bunch of untouched cells individually. If a bin was touched, you then check all the cells within that bin as now.

Reply:
The second point was responded to in https://github.com/OpenRA/OpenRA/pull/17383#issuecomment-558609000 - but I do still think we should actually test this before deciding.

Reply:
Regarding the first point: Adding and removing cells to HashSets was a major performance drain in the previous implementation, so I have aimed to avoid that as much as possible.

Also this being the most expensive trait to tick at the moment shouldn't be that concerning because it is a per-player tick that replaces some logic that would otherwise be done in per-unit ticks.

I am aware that this method sacrifices some performance in the best-case scenario in exchange for better performance in more straining scenarios. That is a good trade IMO because obviously performance matters more in more straining scenarios.

Reply:
In any case, if improvements are still possible here it is IMO reasonable to defer them to a followup PR so long as we aren't regressing our current awful baseline case.